Change to django management command

David-Wobrock opened the pull request:
Turn the linter into a Django management command.
* Simplify migration handling and discovery (use Django internals)
* Speed up the "get_sql" for each migration by doing an internal `call_command`
* Fix cache with different databases
* Simplify tests

To the reviewers, I strongly recommend a commit by commit review. That should ease the task.
I would also be grateful if you could test this version of the linter in your code base:
* add the linter to your `INSTALLED_APPS`
* change the call to a management call `lintmigrations` (either from the command line `python manage.py lintmigrations` or in a Django code `call_command("lintmigrations")`

Fixes #57
Fixes #22

flixx commented:
Hi @David-Wobrock - cool thanks!
we will need some time to review that.

flixx commented:
@David-Wobrock I tested this on our codebase and have to say again: Great work! I can verify that the linter now runs significantly faster.

I made 2 imports that you made relative to make it easier to do drop-in testing. (Just like in the rest of the codebase).

Also I created 2 more PRs based on this one. Would be cool to also have those in 1.0.0.

So approval from me. @mes3yd want to do a code review?

David-Wobrock commented:
Thanks for the review @flixx 
I merged the 2 PRs, I updated the changelog, added a test for the bug you found and added support for Django 2.2 while I was at it.

We can wait a few days if other potential reviewers are busy, but I would be confident to merge the PR as it is and release it :)

tuky commented on django_migration_linter/migration_linter.py:
```diff
@@ -61,51 +63,80 @@ def __init__(self, project_path, **kwargs):
+        md5hash = self.get_migration_hash(app_label, migration_name)
 
-        if self.should_ignore_migration(app_name, migration_name):
+        if self.should_ignore_migration(app_label, migration_name):
```
as mentioned in https://github.com/3YOURMIND/django-migration-linter/issues/57#issuecomment-491286030 i would suggest to provide the `migration` to this method and make it check, if has an operation that `isinstanceof` `IgnoreMigration`. This would skip the later time consuming `get_sql` and `analyse_sql_statements` steps.

tuky commented on django_migration_linter/migration_linter.py:
```diff
@@ -220,132 +220,24 @@ def _gather_migrations_git(self, git_commit_id):
+            if app_label not in DJANGO_APPS_WITH_MIGRATIONS:
+                yield migration
+
+    def should_ignore_migration(self, app_label, migration_name):
```
```suggestion
    def should_ignore_migration(self, migration):
        app_label = migration.app_label
        migration_name = migration.name
```

tuky commented on django_migration_linter/migration_linter.py:
```diff
@@ -220,132 +220,24 @@ def _gather_migrations_git(self, git_commit_id):
-            (self.include_apps and app_name not in self.include_apps)
-            or (self.exclude_apps and app_name in self.exclude_apps)
+            (self.include_apps and app_label not in self.include_apps)
+            or (self.exclude_apps and app_label in self.exclude_apps)
```
```suggestion
            or (self.exclude_apps and app_label in self.exclude_apps)
            or (any([isinstance(o, IgnoreMigration) for o in migration.operations]))
```

tuky commented on django_migration_linter/migration_linter.py:
```diff
@@ -13,46 +13,48 @@
-from .migration import Migration
-from .utils import is_directory, is_django_project, clean_bytes_to_str
+from .constants import DEFAULT_CACHE_PATH
+from .utils import clean_bytes_to_str, get_migration_abspath, split_migration_path
```
```suggestion
from .utils import clean_bytes_to_str, get_migration_abspath, split_migration_path
from .operations import IgnoreMigration
```

tuky commented:
Could you include my suggested changes? They are pretty small :-)

flixx commented:
@tuky Thank you, I like your suggestion!
However I would really like to test it with our codebase before merging. Would it be possible for you to open a separate PR for this?

tuky commented:
sure, i will try to create a PR against https://github.com/3YOURMIND/django-migration-linter/tree/feature/major/change-to-django-command

tuky commented:
You can check out https://github.com/3YOURMIND/django-migration-linter/pull/61/files. i hope i made the tests work, too.

David-Wobrock commented:
I left a review on https://github.com/3YOURMIND/django-migration-linter/pull/61

@flixx I'll probably merge this PR this week-end (but without necessarily releasing 1.0.0 right away, depends on the available time I have)
